File: app/app/src/test/java/com/example/hakonsreader/api/RedditApiTest.java
package com.example.hakonsreader.api;

import org.junit.Test;
import org.mockito.Mockito;

public class RedditApiTest {

    // TODO fix test cases for kotlin

    /**
     * Tests that the builder constructor throws an exception if no user agent is given.
     * The user agent must be present to respect Reddit's API rules of setting a custom
     * User-Agent header
     */
    @Test(expected = IllegalStateException.class)
    public void builderNoUserAgent() {
      //  new RedditApi.Builder("", "gerg");
    }

    /**
     * Tests that the builder constructor throws an exception if no client ID is given
     */
    @Test(expected = IllegalStateException.class)
    public void builderNoClientId() {
   //     new RedditApi.Builder("fffff", "");
    }

    /**
     * Tests that the builder constructor throws an exception if no user agent or client ID is given
     */
    @Test(expected = IllegalStateException.class)
    public void builderNoUserAgentOrClientId() {
     //   new RedditApi.Builder("", "");
    }

    /**
     * Tests that trying to get an access token without a callback URL fails.
     * This has to fail as the callback URL is mandatory to make this call
     */
    @Test(expected = IllegalStateException.class)
    public void getAccessTokenNoCallbackUrl() {
        // TODO fix this for the Kotlin version
   //     new RedditApi.Builder("a", "b").build().accessToken().get("", null, null);
    }

}

File: app/app/src/test/java/com/example/hakonsreader/api/RedditApiTest.kt
package com.example.hakonsreader.api

import org.junit.Test

class RedditApiTest {

    /**
     * Tests that the constructor throws an exception when the userAgent parameter is empty
     */
    @Test(expected = IllegalStateException::class)
    fun noUserAgent() {
        RedditApi.create(userAgent = "", clientId = "hahuierjioogj")
    }

    /**
     * Tests that the constructor throws an exception when the clientId parameter is empty
     */
    @Test(expected = IllegalStateException::class)
    fun noClientId() {
        RedditApi.create(userAgent = "v0.0.0 by u/hakonschia", clientId = "")
    }

    /**
     * Tests that the constructor throws an exception when the userAgent or clientId parameter is empty
     */
    @Test(expected = IllegalStateException::class)
    fun noUserAgentOrClientId() {
        RedditApi.create(userAgent = "", clientId = "")
    }

    /**
     * Tests that calls to [RedditApiImpl.accessToken] throws an exception if the callback URL isn't given
     * when creating the object
     */
    @Test(expected = IllegalStateException::class)
    fun getAccessTokenNoCallbackUrl() {
        // Test with empty callback URL
        RedditApi.create(userAgent = "a", clientId = "b").accessToken()
    }

    /**
     * Tests that calls to [RedditApiImpl.accessToken] throws an exception if the callback URL is given
     * but is empty when creating the object
     */
    @Test(expected = IllegalStateException::class)
    fun getAccessTokenEmptyCallbackUrl() {
        // Test with empty callback URL
        RedditApi.create(userAgent = "a", clientId = "b", callbackUrl = "").accessToken()
    }
}